daffodil-lib/src/main/scala/org/apache/daffodil/util/Validators.scala

```diff
@@ -0,0 +1,71 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.daffodil.util
+
+import java.util.ServiceLoader
+
+import org.apache.daffodil.api.CompiledValidator
+import org.apache.daffodil.api.Validator
+import org.apache.daffodil.api.Validator.CompilerOps.CheckArgs
+import org.apache.daffodil.api.ValidatorNotFoundException
+
+import scala.collection.JavaConverters._
+
+object Validators {
+  val default: Validator = new DefaultValidatorSPIProvider
+  def compiler(): ValidatorCompiler = new Compiler
+
+  trait ValidatorCompiler {
+    def find(name: String): Option[CompiledValidator]
+    def isRegistered(name: String): Boolean
+    def compile(name: String, args: Validator.ArgumentList): CompiledValidator
+  }
+
+  private class Compiler extends ValidatorCompiler {
+    private var compiled = Map.empty[String, CompiledValidator]
+    private lazy val impls: Map[String, Validator] =
+      ServiceLoader
+        .load(classOf[Validator])
+        .iterator()
+        .asScala
+        .map(v => v.name() -> v)
+        .toMap
+
```

Comment:
Can you describe the motivation for using a ServiceLoader for providing validators rather than just  providing an API where a user can provide validators (that follow the Validator API, and meeting requirements like thread safe and Serializable)? I feel like SPI things are useful when you just want to swap different mechanisms without any code/configuration changes, but it seems like that's not really the case here. Swapping different validation mechanisms requires speciying the validation name somewhere (I think) and somehow providing args, so swapping in a new validation mechamisnm still requires code changes.

Reply:
Sure.  The primary motivation was to drop validators into binary distributions of daffodil, both for cli and artifacts.  The sample application demonstrates the process and has installation instructions in the readme.

> I feel like SPI things are useful when you just want to swap different mechanisms without any code/configuration changes, but it seems like that's not really the case here.

The clearest way to describe it is as a plugin framework.  You can define an interface in your software and at runtime collect all instances of that type that exist on the classpath and use them for whatever.  No need for a client to register anything other than providing a text file in META-INF.  And also no need for the client to package up Daffodil in their binary distribution.

> Swapping different validation mechanisms requires speciying the validation name somewhere (I think) and somehow providing args

The name is provided by the implementation, see the Validator trait.  The args are provided either programmatically or via the CLI.  My primary motivation here was supplementing CLI usage of the binary distribution of Daffodil without recompiling Daffodil, that was what the first implementation was based on, which led to some of the feedback that rightly pointed out CLI concepts were leaking into the API.  Definitely want to make it easy to use programmatically, and I think thats still a bit of a WIP that needs some polish yet.

> so swapping in a new validation mechamisnm still requires code changes.

Only on the client side, Daffodil defines an API, finds things through ServiceLoader, and handles them all abstractly.

Take a look at the sample application that went with the original reference impl.  Its drifted some as I have tried to adjust to the feedback, but the concepts are still the same.

https://github.com/ctc-oss/daffodil-schematron-validator

Reply:
I guess then I'm interested with how the API will get fleshed out. Based on how the CLI uses validators, it feels like it's going to end up as something like this:

```scala
dp.withValidator(Validator.Custom("sch", "key1=value1 key2=value2"))
```
Which feels somewhat akward to me.

Yes, you get the benefit that you don't have to recompile your app when you want to change the validator, but it's sitll not just drop in a jar and go. You still need to give your app some mechanism to provide the validator name to use and provide configuration options. And the configuration options need to be representable as key/value pairs which feels limiting. And every validator would likely have differnt key names. So even though you don't need to recompile, you still need some mechamism to pick and configure different avilable validators.

To me this kind of plugability seems useful for the CLI and those limitations/awkwardness feel reasonable since it is a CLI and not an pipeline orchestration engine. But as an API user, I would rather do something like:

```scala
val validator = configureAndCompileValidatorOfChoice()
dp.withValidation(validator)
```

If my app needs plugablity and the validators have simple configuration options, then I could always implement it much like you have using SPI and key/value pairs. But if my app has complex Validator's that need some configuration other then key/value pair strings (e.g. an open socket to connect to a rules engine, dynamically generated validation rules) where the preferable configuration is via code, then we still have that option.

Note that this feels different to me than Daffodil's UserDefinedFunctions which also use SPI. But UDFs feel different because they literally are drop and go. There's no configuration that needs to be done or selection of which ones to use. You just drop in a UDF jar on the class path and if a schema uses a UDF by that name then we use it as is.

Reply:
> dp.withValidator(Validator.Custom("sch", "key1=value1 key2=value2"))

Nah, it never would have ended up like that. I should have focused more on API before submitting PR, but here we are.

Ultimately I want to get as close to type safety and convenience as we can.  There was some work towards this in the commits earlier today.

> val validator = configureAndCompileValidatorOfChoice()
> dp.withValidation(validator)

Yeah, I think we are heading in that direction now.  Are you tracking @mbeckerle's compiled approach?

> Note that this feels different to me than Daffodil's UserDefinedFunctions which also use SPI. But UDFs feel different because they literally are drop and go. There's no configuration that needs to be done or selection of which ones to use. You just drop in a UDF jar on the class path and if a schema uses a UDF by that name then we use it as is.

Yeah, realizing I needed to handle arguments was a bit of a shock that came after thinking the implementation was simple :) and then the shock of needing multiple arguments followed that.

Have an idea Ill post at the top level that just came to me when discussing this.  Thanks

Reply:
Sounds good! I haven't looked in detail about @mbeckerle's approach, I'll take a more detailed look tomorrow.